src/Chainweb/Cut/Create.hs

```diff
@@ -0,0 +1,442 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Module: Chainweb.Cut.Create
+-- Copyright: Copyright © 2020 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>
+-- Stability: experimental
+--
+-- The steps for extending a Cut with a new Block are usually as follows:
+--
+-- 1. Select a chain for the new Block header
+-- 2. Call 'getCutExtension'. If the result is 'Nothing' the chain is
+--    blocked. In which case another chain must be choosen.
+-- 3. Call pact service to create new payload for the new block
+-- 4. Assemble preliminary work header, with preliminary time and nonce value
+--    (time should be current time, nonce should be 0) and send work header
+--    to the miner.
+-- 5. Receive resolved header (with final nonce and creation time) from miner
+-- 6. Create new header by computing the merkle hash for the resolved header
+-- 7. Extend cut with new header by calling 'tryMonotonicCutExtension'
+-- 8. Create new cut hashes with the new header and payload attached.
+-- 9. Submit cut hashes (along with attached new header and payload) to cut
+--    validation pipeline.
+--
+--
+module Chainweb.Cut.Create
```

Comment:
Perhaps `Mine` instead of `Create`? It would seem cut creation occurs (in production) strictly in the context of mining, and it would motivate including mining-specific types like `WorkHeader` and `SolvedWork` in here.

Reply:
This code is meant to be the first draft of what should become the consensus side of API for mining code.

The mining code should not need to now about the internal of block headers. In particular it shouldn't have to deal with the construction of Merkel trees and Cut braidings. The knowledge of the Merkel trees and braidings should be limited to a small core of the consensus code.

The mining API only receives an `Work` item and all it knows is 

1. where to inject the nonce,
2. where to inject the creation time,
3. how to compute the POW hash,
4. compare the POW hash with the target. 

Otherwise the work item should be opaque to the miner. Ideally, I `Work` item would be just 

1. Initial Nonce,
2. Initial time,
3. some random bytes (a cryptographic hash of the remaining data in the header), and
4. possibly some feature flags.

At this point the (3.) part is spread out in between the other parts -- something we may want to fix at some point.

When a miner has found a valid nonce, it returns the result as `SolvedWork` item.

The code in the module will change in upcoming PRs.